cocotb/bfmgen.py

```diff
@@ -0,0 +1,346 @@
+###############################################################################
+# Copyright cocotb contributors
+# Licensed under the Revised BSD License, see LICENSE for details.
+# SPDX-License-Identifier: BSD-3-Clause
+###############################################################################
+"""Generates the HDL BFM shell based on loaded Python BFM classes"""
+
+import os
+import argparse
+import importlib
+from string import Template
+
+import cocotb
+from cocotb.bfms import BfmMgr
+
+
+def bfm_load_modules(module_l):
+    for m in module_l:
+        importlib.import_module(m)
+
+def process_template_vl(template, info):
+    """Process a single BFM-HDL template file string (template),
+    substituting generated interface tasks and task-call demux
+    based on BFM task declarations (info). Returns complete
+    BFM Verilog module as a string"""
+
+    t = Template(template)
+
+    bfm_import_calls = ""
+    for i,imp in enumerate(info.import_info):
+        bfm_import_calls += "              " + str(i) + ": begin\n"
+        bfm_import_calls += "                  " + imp.T.__name__  + "(\n"
+        for pi,p in enumerate(imp.signature):
+            if p.ptype.s:
+                bfm_import_calls += "                      $cocotb_bfm_get_param_i32(bfm_id)"
+            else:
+                bfm_import_calls += "                      $cocotb_bfm_get_param_ui32(bfm_id)"
+
+            if pi+1 < len(imp.signature):
+                bfm_import_calls += ","
+            bfm_import_calls += "\n"
+        bfm_import_calls += "                      );\n"
+        bfm_import_calls += "              end\n"
+
+    bfm_export_tasks = ""
+    for i,exp in enumerate(info.export_info):
+        bfm_export_tasks += "    task " + exp.T.__name__
+
+        if len(exp.signature) > 0:
+            bfm_export_tasks += "("
+            for j,p in enumerate(exp.signature):
+                bfm_export_tasks += p.ptype.vl_type() + " " + p.pname
+                if j+1 < len(exp.signature):
+                    bfm_export_tasks += ", "
+            bfm_export_tasks += ");\n"
+        else:
+            bfm_export_tasks += ";\n"
+        bfm_export_tasks += "    begin\n"
+        bfm_export_tasks += "        $cocotb_bfm_begin_msg(bfm_id, " + str(i) + ");\n"
+        for p in exp.signature:
+            if p.ptype.s:
+                bfm_export_tasks += "        $cocotb_bfm_add_param_si(bfm_id, " + p.pname + ");\n"
+            else:
+                bfm_export_tasks += "        $cocotb_bfm_add_param_ui(bfm_id, " + p.pname + ");\n"
+
+        bfm_export_tasks += "        $cocotb_bfm_end_msg(bfm_id);\n"
```

Comment:
It's tempting to make this a single call, `$cocotb_bfm_send_msg(bfm_id, i, p1, p2, p3, ...)`, as:

```
        bfm_export_tasks += "        $cocotb_bfm_begin_msg(bfm_id, {}{});\n".format(
            i,
            "".join(", {}".format(p.pname)) for p in exp.signature)
        )
```
and have `$cocotb_bfm_begin_msg` work out the signed vs unsigned from the value of `i`.